py3status/modules/battery_level.py

```diff
@@ -168,13 +168,18 @@ def post_config_hook(self):
             raise NameError("Invalid measurement mode")
 
     def battery_level(self):
-        if not os.listdir(self.sys_battery_path):
+        if self.measurement_mode == "acpi":
+            battery_list = self._extract_battery_information_from_acpi()
+        else:
+            battery_list = self._extract_battery_information_from_sys()
+
+        if not battery_list:
```

Comment:
Can you put this in `post_config_hook` instead similar to `arch_updates`?
```python
...
        elif self.measurement_mode == "acpi":
            self._get_battery_info = self._extract_battery_information_from_acpi
...
battery_list = self._get_battery_info()
if not battery_list:
```

Reply:
Wouldn't this make the code to never refresh battery information, even if a new one is connected/disconnected?

Reply:
>It also allows _extract_battery_information_from_acpi() to run even when /sys/class/power_supply/ does not exists, however this seems to be fine since ACPI can use other methods to check the battery level.

I don't see this in code. Can you point it out? I see that we use acpi or sys based on `measurement_mode` and `measurement_mode` does not change after getting assigned.

>Wouldn't this make the code to never refresh battery information, even if a new one is connected/disconnected?

Shouldn't. Does not look like that. Notice the missing `()` calls. We assign function once then call that repeatedly instead of checking if statement + call repeatedly.

Unrelated: Cleaning up `measurement_mode` messages too for log + error.
```diff
diff --git a/py3status/modules/battery_level.py b/py3status/modules/battery_level.py
index 77a8588..08bd55d 100644
--- a/py3status/modules/battery_level.py
+++ b/py3status/modules/battery_level.py
@@ -163,16 +163,18 @@ class Py3status:
             elif self.py3.check_commands(["acpi"]):
                 self.measurement_mode = "acpi"
 
-        self.py3.log("Measurement mode: " + self.measurement_mode)
-        if self.measurement_mode != "acpi" and self.measurement_mode != "sys":
-            raise NameError("Invalid measurement mode")
-
-    def battery_level(self):
-        if self.measurement_mode == "acpi":
-            battery_list = self._extract_battery_information_from_acpi()
+        msg = "measurement_mode `{}`".format(self.measurement_mode)
+        if self.measurement_mode == "sys":
+            self.get_battery_info = self._extract_battery_info_from_sys
+        elif self.measurement_mode == "acpi":
+            self.get_battery_info = self._extract_battery_info_from_acpi
         else:
-            battery_list = self._extract_battery_information_from_sys()
+            msg = "invalid {}".format(msg)
+            raise NameError(msg)
+        self.py3.log("selected {}".format(msg))
 
+    def battery_level(self):
+        battery_list = self.get_battery_info()
         if not battery_list:
             return {
                 "full_text": "",
@@ -211,7 +213,7 @@ class Py3status:
         if message:
             self.py3.notify_user(message, "info")
 
-    def _extract_battery_information_from_acpi(self):
+    def _extract_battery_info_from_acpi(self):
         """
         Get the battery info from acpi
 
@@ -262,7 +264,7 @@ class Py3status:
 
         return [_parse_battery_info(battery) for battery in acpi_list]
 
-    def _extract_battery_information_from_sys(self):
+    def _extract_battery_info_from_sys(self):
         """
         Extract the percent charged, charging state, time remaining,
         and capacity for a battery, using Linux's kernel /sys interface

```

Reply:
> I don't see this in code. Can you point it out? I see that we use acpi or sys based on measurement_mode and measurement_mode does not change after getting assigned.

Since the check for file was in `battery_level()` call, even if the mode was set to `acpi` you would need to have something in `/sys/class/power_supply/`. `acpi`, AFAIK, can work with the old `proc` system and even in non-Linux systems.

> Shouldn't. Does not look like that. Notice the missing () calls. We assign function once then call that repeatedly instead of checking if statement + call repeatedly.

Makes sense, I will do it :+1: .

Reply:
@lasers I applied your patch, rebased and tested in both my Desktop and Laptop. Seems fine :+1: .

Reply:
Can you fix my stupid mistake?
```diff
diff --git a/py3status/modules/battery_level.py b/py3status/modules/battery_level.py
index 08bd55d..1bbf3a6 100644
--- a/py3status/modules/battery_level.py
+++ b/py3status/modules/battery_level.py
@@ -169,8 +169,7 @@ class Py3status:
         elif self.measurement_mode == "acpi":
             self.get_battery_info = self._extract_battery_info_from_acpi
         else:
-            msg = "invalid {}".format(msg)
-            raise NameError(msg)
+            raise NameError("invalid {}".format(msg))
         self.py3.log("selected {}".format(msg))
 
     def battery_level(self):
```
I borrowed PS4 controller. Am trying to set it up right now. :-)

Reply:
> Can you fix my stupid mistake?

Aye. Applied :+1: .